### PR TITLE
Prevent dblclick event being handled by view tab when only clicked once (#24)

### PR DESCRIPTION
**Pull request by mofogasy**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit follows our [guidelines](https://github.com/SchweizerischeBundesbahnen/scion-workbench/blob/master/CONTRIBUTING.md#contribution)
- [ ] Tests for the changes have been added (required for routing or view grid related issues)
- [ ] Public API has been documented

## PR Type
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Issue
Issue Number: 24

## Does this PR introduce a breaking change?
```
[ ] Yes
[x ] No
```

## Other information

### CODE REVIEW

**Comment by danielwiehl** [comment]
Matthias, thanks again for your pull request, fantastic!

If ever possible, we try to avoid using `setTimeout`, unless it is inevitable (e.g. post construct initializer or similar if any). Often, it is difficult to guess correct delays, especially to not loose built-in behavior like acceleration or deceleration.

I suggest fixing the problem at its very source where the event is fired. Why not simply attach a `dblclick` listener to the close button to prevent further propagation? That way, the event does not bubble and would not be handled by the view tab.

In its simplest form, I suggest to change `view-tab.component.html` to something like this:

```
<button class="close wb-icons"
        ...
        (dblclick)="$event.stopPropagation()"
        (click)="onClose($event)">
</button>
```

**Comment by mofogasy** [comment]
Hello Daniel,

Thanks for your reply. You're right regarding the use of setTimeout. Unfortunately in this case I don't see an alternative. Your solution will work if the tabs have the same size. But if they sizes are different, then not the close button will get the dblclick event, but the tab itself, in which case we somehow have to track if it was clicked before or not.

Cheers,

Matthias

**Comment by danielwiehl** [comment]
Gotcha, good point. However, I still don't like the idea of using a timer here. I think, there is a better way. 

We both agree that maximization should not occur if closing views quickly. As you suggested, we need some kind of state to enable or disable that feature. The state could be `mouseenter`, `mousemove` and `mouseleave`, meaning that the feature is only enabled after `mousenter` and `mousemove` occurred. Typically, when closing view tabs quickly, the user does not move the mouse.

I did some coding and came up with the following snippet.

Please note to listen for `mousemove` events outside of the Angular zone to not change detect the application too excessively. Also, I put state management and listening for double click events into the same method, so it is simpler to pick up the idea.

```
  /**
   * Listens for 'dblclick' events to maximize or minimize the application main content, unless closing views quickly.
   */
  private installMaximizeListener(zone: NgZone): void {
    // Install listeners outside the Angular zone to not change detect the application upon every mouse event.
    zone.runOutsideAngular(() => {
      let enabled = false;

      merge(fromEvent(this._host, 'mouseenter'), fromEvent(this._host, 'mousemove'), fromEvent(this._host, 'mouseleave'))
        .pipe(takeUntil(this._destroy$))
        .subscribe((event: Event) => {
          enabled = (event.type === 'mousemove');
        });

      fromEvent(this._host, 'dblclick')
        .pipe(takeUntil(this._destroy$))
        .subscribe((event: Event) => {
          enabled && zone.run(() => this._workbenchLayout.toggleMaximized());
          event.stopPropagation(); // prevent toggling in view tab bar
        });
    });
  }
```

**Comment by mofogasy** [comment]
Nice solution! I find it quite readable so I left this fix in the method you suggested instead of splitting it up.
There is still a tiny problem: after maximizing, the mouse can be on another tab. Double-click won't work then because the mouse wasn't moved. I tried a couple of things to eliminate this issue as well, but unfortunately I wasn't successful. If I get a sudden idea I'll tell you, in the meantime I think we can live with this solution.

File: projects/scion/workbench/src/lib/view-part/view-tab/view-tab.component.ts [hunk 1]
```diff
@@ -41,8 +43,14 @@ export class ViewTabComponent {
...
               private _workbenchLayout: WorkbenchLayoutService,
               private _viewport: SciViewportComponent,
-              private _viewPartService: WorkbenchViewPartService) {
+              private _viewPartService: WorkbenchViewPartService,
+              zone: NgZone) {
     this._host = host.nativeElement;
+    this.installMaximizeListener(zone);
+  }
+
+  ngOnDestroy(): void {
```
Comment (danielwiehl) [inline projects/scion/workbench/src/lib/view-part/view-tab/view-tab.component.ts]: Please specify `public` access modifier and move this method to the end of the file.

```suggestion
  public ngOnDestroy(): void {
```

File: projects/scion/workbench/src/lib/view-part/view-tab/view-tab.component.ts [hunk 2]
```diff
@@ -153,4 +155,30 @@ export class ViewTabComponent {
...
+        });
+
+      fromEvent(this._host, 'dblclick')
+        .pipe(
+          filter(() => enabled),
+          takeUntil(this._destroy$)
+        )
+        .subscribe((event: Event) => {
+          zone.run(() => this._workbenchLayout.toggleMaximized());
+          event.stopPropagation(); // prevent toggling in view tab bar
```
Comment (danielwiehl) [inline projects/scion/workbench/src/lib/view-part/view-tab/view-tab.component.ts]: Please change the comment to something like this: 
// prevent \`ViewPartBarComponent\` handling the dblclick event which would undo maximization/minimization

File: projects/scion/workbench/src/lib/view-part/view-tab/view-tab.component.ts (hunk 1 above)
Comment (mofogasy) [reply projects/scion/workbench/src/lib/view-part/view-tab/view-tab.component.ts]: done

File: projects/scion/workbench/src/lib/view-part/view-tab/view-tab.component.ts (hunk 2 above)
Comment (mofogasy) [reply projects/scion/workbench/src/lib/view-part/view-tab/view-tab.component.ts]: done

**Pull request merged**